File: sdk/samples/05_security/python/authentication.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Authentication Sample - Demonstrates PKI-based participant authentication concepts

This sample shows DDS Security authentication concepts:
- Certificate-based identity (X.509)
- CA trust chain validation
- Mutual authentication between participants

Key concepts:
- Identity Certificate and Private Key
- Certificate Authority (CA) for trust
- Authentication plugin configuration

Note: Security features are not yet fully implemented in HDDS.
      This sample demonstrates the concepts while using the basic API.

Prerequisites:
  Generate certificates using: ../certs/generate_certs.sh

NOTE: CONCEPT DEMO - This sample demonstrates the APPLICATION PATTERN for DDS Security Authentication.
The native DDS Security Authentication API is not yet exported to the C/C++/Python SDK.
This sample uses standard participant/writer/reader API to show the concept.
"""

import sys
import time
from dataclasses import dataclass
from pathlib import Path

# Add SDK to path
sys.path.insert(0, '../../../python')

import hdds


@dataclass
class AuthenticationConfig:
    """Authentication configuration for DDS Security"""
    identity_ca: str       # CA certificate path
    identity_cert: str     # Participant certificate path
    private_key: str       # Private key path
    password: str = ""     # Private key password (optional)


def get_certs_dir() -> Path:
    """Get path to certificates directory"""
    return Path(__file__).parent.parent / "certs"


def print_cert_info(label: str, path: Path):
    """Print certificate file info"""
    status = "[OK]" if path.exists() else "[NOT FOUND]"
    print(f"  {label}: {path} {status}")


def run_publisher(participant, participant_name):
    """Run publisher sending authenticated messages."""
    print("Creating writer for SecureData topic...")
    writer = participant.create_writer("SecureData")

    print("\n--- Sending Authenticated Messages ---\n")

    for msg_count in range(1, 6):
        message = f"Authenticated message #{msg_count} from {participant_name}"
        print(f"[SEND] {message}")
        writer.write(message.encode('utf-8'))
        print("       (message signed with participant identity)")
        time.sleep(2)

    print("\nDone sending.")


def run_subscriber(participant):
    """Run subscriber receiving authenticated messages."""
    print("Creating reader for SecureData topic...")
    reader = participant.create_reader("SecureData")

    # Create waitset for efficient waiting
    waitset = hdds.WaitSet()
    cond = reader.get_status_condition()
    waitset.attach(cond)

    print("\n--- Waiting for Authenticated Messages ---\n")
    print("Run a publisher with:")
    print("  python authentication.py pub [ParticipantName]\n")

    received = 0
    max_receive = 10

    while received < max_receive:
        if waitset.wait(timeout=5.0):
            while True:
                data = reader.take()
                if data is None:
                    break
                message = data.decode('utf-8')
                print(f"[RECV] {message}")
                print("       (sender identity verified)")
                received += 1
        else:
            print("  (waiting for authenticated peers...)")

    print("\nDone receiving.")


def main():
    print("=== HDDS Authentication Sample ===\n")
    print("NOTE: CONCEPT DEMO - Native DDS Security Authentication API not yet in SDK.")
    print("      Using standard pub/sub API to demonstrate the pattern.\n")

    # Parse arguments
    is_publisher = len(sys.argv) > 1 and sys.argv[1].lower() in ('pub', 'publisher', '-p')
    participant_name = sys.argv[2] if len(sys.argv) > 2 else "Participant1"

    certs_dir = get_certs_dir()

    # Configure authentication (conceptual - security not yet implemented)
    auth_config = AuthenticationConfig(
        identity_ca=str(certs_dir / "ca_cert.pem"),
        identity_cert=str(certs_dir / f"{participant_name}_cert.pem"),
        private_key=str(certs_dir / f"{participant_name}_key.pem"),
        password=""
    )

    print("Security Configuration:")
    print_cert_info("CA Certificate", Path(auth_config.identity_ca))
    print_cert_info("Identity Cert ", Path(auth_config.identity_cert))
    print_cert_info("Private Key   ", Path(auth_config.private_key))
    print()

    print("--- DDS Security Authentication ---")
    print("Authentication uses X.509 PKI:")
    print("1. Each participant has an identity certificate")
    print("2. Certificates are signed by a trusted CA")
    print("3. Participants validate each other's certificates")
    print("4. Only authenticated participants can communicate\n")

    # Initialize logging
    hdds.logging.init(hdds.LogLevel.INFO)

    # Create participant
    print(f"Creating DomainParticipant '{participant_name}'...")
    participant = hdds.Participant(participant_name)
    print("[OK] Participant created")

    # Display authentication status (conceptual)
    print("\nAuthentication Status (conceptual):")
    print(f"  Authenticated: YES")
    print(f"  Local Identity: CN={participant_name},O=HDDS,C=US")
    print(f"  Status: AUTHENTICATED\n")

    try:
        if is_publisher:
            run_publisher(participant, participant_name)
        else:
            run_subscriber(participant)
    except KeyboardInterrupt:
        print("\nInterrupted.")

    # Show authentication summary
    print("\n--- Authentication Summary ---")
    print(f"This participant: {participant_name}")
    print("Authentication: SUCCESS (conceptual)")
    print("\nNote: Full security features are not yet implemented.")
    print("      This sample demonstrates authentication concepts")
    print("      while using the basic HDDS API for communication.")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/samples/05_security/rust/src/bin/authentication.rs
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

//! # HDDS Sample: Authentication
//!
//! Demonstrates **PKI-based authentication** - X.509 certificate-based identity
//! verification for DDS participants.
//!
//! ## Authentication Flow
//!
//! ```text
//! ┌──────────────────────────────────────────────────────────────────────┐
//! │                    Certificate Authority (CA)                        │
//! │                    ┌─────────────────────┐                           │
//! │                    │  ca_cert.pem        │                           │
//! │                    │  (trusted root)     │                           │
//! │                    └──────────┬──────────┘                           │
//! │                               │ signs                                │
//! │              ┌────────────────┼────────────────┐                     │
//! │              ▼                                 ▼                     │
//! │   ┌─────────────────────┐          ┌─────────────────────┐          │
//! │   │ participant1_cert   │          │ participant2_cert   │          │
//! │   │ participant1_key    │          │ participant2_key    │          │
//! │   └──────────┬──────────┘          └──────────┬──────────┘          │
//! │              │                                 │                     │
//! │              ▼                                 ▼                     │
//! │   ┌─────────────────────┐          ┌─────────────────────┐          │
//! │   │  Participant 1      │◄────────►│  Participant 2      │          │
//! │   │  (authenticated)    │  mutual  │  (authenticated)    │          │
//! │   └─────────────────────┘   auth   └─────────────────────┘          │
//! └──────────────────────────────────────────────────────────────────────┘
//! ```
//!
//! ## Certificate Components
//!
//! | File                 | Purpose                        |
//! |----------------------|--------------------------------|
//! | ca_cert.pem          | Certificate Authority root     |
//! | participant_cert.pem | Participant's identity cert    |
//! | participant_key.pem  | Participant's private key      |
//!
//! ## X.509 Subject Name
//!
//! ```text
//! CN=Participant1,O=HDDS,C=US
//! │              │         │
//! │              │         └── Country
//! │              └── Organization
//! └── Common Name (identity)
//! ```
//!
//! ## Use Cases
//!
//! - **Identity verification**: Ensure only authorized nodes join
//! - **Mutual authentication**: Both parties verify each other
//! - **Audit trail**: Track which certificate accessed what
//!
//! ## Running the Sample
//!
//! ```bash
//! # Terminal 1 - First participant (publisher)
//! cargo run --bin authentication
//!
//! # Terminal 2 - Second participant (subscriber)
//! cargo run --bin authentication -- Participant2
//! ```
//!
//! **NOTE: CONCEPT DEMO** - This sample demonstrates the APPLICATION PATTERN for DDS Security Authentication.
//! The native DDS Security Authentication API is not yet exported to the SDK.
//! This sample uses standard participant/writer/reader API to show the concept.

use std::env;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

// Include generated types
#[allow(dead_code)]
mod generated {
    include!("../../generated/security_types.rs");
}

use generated::SecureMessage;

/// Authentication configuration (for documentation purposes)
#[derive(Debug, Clone)]
#[allow(dead_code)]
struct AuthenticationConfig {
    identity_ca: PathBuf,     // CA certificate path
    identity_cert: PathBuf,   // Participant certificate path
    private_key: PathBuf,     // Private key path
    password: Option<String>, // Private key password
}

/// Authentication status
#[derive(Debug)]
struct AuthStatus {
    authenticated: bool,
    peer_identity: String,
    status_message: String,
}

fn get_certs_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../certs")
}

fn print_cert_info(label: &str, path: &Path) {
    let status = if path.exists() { "[OK]" } else { "[NOT FOUND]" };
    println!("  {}: {} {}", label, path.display(), status);
}

fn print_security_concepts() {
    println!("--- DDS Security Authentication ---");
    println!("Authentication uses X.509 PKI:");
    println!("1. Each participant has an identity certificate");
    println!("2. Certificates are signed by a trusted CA");
    println!("3. Participants validate each other's certificates");
    println!("4. Only authenticated participants can communicate");
    println!();
}

fn run_publisher(participant: &Arc<hdds::Participant>, name: &str) -> Result<(), hdds::Error> {
    println!("Creating authenticated writer...");
    let writer = participant.create_raw_writer("SecureAuthTopic", None)?;

    println!("[OK] Secure DataWriter created");
    println!();
    println!("--- Sending Authenticated Messages ---");
    println!("Run another instance with different identity:");
    println!("  cargo run --bin authentication -- Participant2");
    println!();

    for i in 1..=5 {
        let msg = SecureMessage::new(name, format!("Authenticated message #{}", i), i);
        let data = msg.serialize();

        writer.write_raw(&data)?;
        println!("[SEND] {} (seq={})", msg.payload, msg.sequence);
        println!("       From authenticated sender: {}", msg.sender_id);

        thread::sleep(Duration::from_secs(2));
    }

    println!();
    println!("Done sending authenticated messages.");
    Ok(())
}

fn run_subscriber(participant: &Arc<hdds::Participant>, name: &str) -> Result<(), hdds::Error> {
    println!("Creating authenticated reader...");
    let reader = participant.create_raw_reader("SecureAuthTopic", None)?;

    println!("[OK] Secure DataReader created");
    println!();
    println!("--- Waiting for Authenticated Messages ---");
    println!("Receiver identity: {}", name);
    println!();

    let mut received = 0;
    let mut attempts = 0;
    while received < 5 && attempts < 30 {
        let samples = reader.try_take_raw()?;
        if samples.is_empty() {
            println!("  (waiting for authenticated peers...)");
            thread::sleep(Duration::from_secs(1));
            attempts += 1;
        } else {
            for sample in samples {
                match SecureMessage::deserialize(&sample.payload) {
                    Ok((msg, _)) => {
                        println!("[RECV] {} (seq={})", msg.payload, msg.sequence);
                        println!("       Verified sender: {}", msg.sender_id);
                        received += 1;
                    }
                    Err(e) => {
                        eprintln!("  Deserialize error: {}", e);
                    }
                }
            }
        }
    }

    println!();
    println!("Done receiving authenticated messages.");
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== HDDS Authentication Sample ===");
    println!();
    println!("NOTE: CONCEPT DEMO - Native DDS Security Authentication API not yet in SDK.");
    println!("      Using standard pub/sub API to demonstrate the pattern.\n");

    let args: Vec<String> = env::args().collect();
    let participant_name = args.get(1).map(|s| s.as_str()).unwrap_or("Participant1");
    let is_publisher = args
        .get(2)
        .map(|s| s == "pub")
        .unwrap_or(participant_name == "Participant1");

    let certs_dir = get_certs_dir();

    // Configure authentication (conceptual - would be used with DDS Security)
    let auth_config = AuthenticationConfig {
        identity_ca: certs_dir.join("ca_cert.pem"),
        identity_cert: certs_dir.join(format!("{}_cert.pem", participant_name)),
        private_key: certs_dir.join(format!("{}_key.pem", participant_name)),
        password: None,
    };

    println!("Security Configuration:");
    print_cert_info("CA Certificate", &auth_config.identity_ca);
    print_cert_info("Identity Cert ", &auth_config.identity_cert);
    print_cert_info("Private Key   ", &auth_config.private_key);
    println!();

    print_security_concepts();

    // Create participant
    println!(
        "Creating secure DomainParticipant '{}'...",
        participant_name
    );
    let participant = hdds::Participant::builder(participant_name)
        .domain_id(0)
        .build()?;
    println!("[OK] Participant created");

    // Simulated authentication status
    let status = AuthStatus {
        authenticated: true,
        peer_identity: format!("CN={},O=HDDS,C=US", participant_name),
        status_message: "AUTHENTICATED".to_string(),
    };

    println!();
    println!("Authentication Status:");
    println!(
        "  Authenticated: {}",
        if status.authenticated { "YES" } else { "NO" }
    );
    println!("  Local Identity: {}", status.peer_identity);
    println!("  Status: {}", status.status_message);
    println!();

    if is_publisher {
        run_publisher(&participant, participant_name)?;
    } else {
        run_subscriber(&participant, participant_name)?;
    }

    // Show authentication summary
    println!();
    println!("--- Authentication Summary ---");
    println!("This participant: {}", participant_name);
    println!(
        "Authentication: {}",
        if status.authenticated {
            "SUCCESS"
        } else {
            "FAILED"
        }
    );
    println!();
    println!("Note: In a production system with DDS Security enabled,");
    println!("      unauthenticated participants would be rejected.");
    println!("      Only peers with valid certificates could join.");

    println!();
    println!("=== Sample Complete ===");
    Ok(())
}
